x-pack/plugins/gis/public/shared/layers/tile_layer.js

```diff
@@ -30,14 +31,52 @@ export class TileLayer extends AbstractLayer {
     return tileLayerDescriptor;
   }
 
+  _tileLoadErrorTracker(map, url) {
+    let tileLoad;
+    map.on('dataloading', ({ tile }) => {
+      if (tile && tile.request) {
+        // If at least one tile loads, endpoint/resource is valid
+        tile.request.onloadend = ({ loaded }) => loaded && (tileLoad = true);
+      }
+    });
+
+    return new Promise((resolve, reject) => {
+      let tileLoadTimer = null;
 
-  syncLayerWithMB(mbMap) {
+      const clearChecks = () => {
+        clearTimeout(tileLoadTimer);
+        map.off('dataloading');
+      };
 
+      tileLoadTimer = setTimeout(() => {
+        if (!tileLoad) {
+          reject(new Error(`Tiles from "${url}" could not be loaded`));
+        } else {
+          resolve();
+        }
+        clearChecks();
+      }, this.TMS_LOAD_TIMEOUT);
+    });
+  }
+
+  async syncLayerWithMB(mbMap) {
     const source = mbMap.getSource(this.getId());
     const layerId = this.getId() + '_raster';
-    if (!source) {
-      const url = this._source.getUrlTemplate();
-      mbMap.addSource(this.getId(), {
+
+    if (source) {
+      return;
+    }
+
+    let url;
+    return new Promise((resolve, reject) => {
```

Comment:
Nothing in this first promise block is async. Why does the first new Promise exist? It really hurts readability to have all of these nested Promises and `.then` callbacks.

This function can be rewritten to the below which is much easier to follow. If this._source.getUrlTemplate or await this._tileLoadErrorTracker(mbMap, url) fail the function will throw an error which will be catch by the caller x-pack/plugins/gis/public/components/map/mb/view.js

```
  async syncLayerWithMB(mbMap) {
    const source = mbMap.getSource(this.getId());
    const layerId = this.getId() + '_raster';

    if (source) {
      return;
    }

    const url = this._source.getUrlTemplate();
    const sourceId = this.getId();
    mbMap.addSource(sourceId, {
      type: 'raster',
      tiles: [url],
      tileSize: 256,
      scheme: 'xyz',
    });

    mbMap.addLayer({
      id: layerId,
      type: 'raster',
      source: sourceId,
      minzoom: 0,
      maxzoom: 22,
    });

    await this._tileLoadErrorTracker(mbMap, url);

    mbMap.setLayoutProperty(layerId, 'visibility', this.isVisible() ? 'visible' : 'none');
    mbMap.setLayerZoomRange(layerId, this._descriptor.minZoom, this._descriptor.maxZoom);
    this._style && this._style.setMBPaintProperties({
      alpha: this.getAlpha(),
      mbMap,
      layerId,
    });
  }
```